api/saml/metadata.py

```diff
@@ -1034,6 +1035,55 @@ def default(self, subject):
         return result
 
 
+class SubjectJSONDecoder(JSONDecoder):
+    """Subject's JSON decoder."""
+
+    def decode(self, raw_subject, _w=WHITESPACE.match):
+        """Decode a JSON document into Subject object.
+
+        :param raw_subject: String containing JSON document
+        :type raw_subject: str
+
+        :param _w: Regular expression used to match white spaces
+        :type _w: RegEx
+
+        :return: Subject object
+        :rtype: api.saml.metadata.Subject
+        """
+        raw_subject = super(SubjectJSONDecoder, self).decode(raw_subject, _w)
+        attribute_statement = None
+        name_id = None
+
+        if 'name_id' in raw_subject:
+            raw_name_id_dict = raw_subject['name_id']
+            raw_name_format = raw_name_id_dict['name_format']
+            raw_name_id = raw_name_id_dict['name_id']
+            raw_name_qualifier = raw_name_id_dict['name_qualifier']
+            raw_sp_name_qualifier = raw_name_id_dict['sp_name_qualifier']
+            name_id = NameID(
+                raw_name_format,
+                raw_name_qualifier,
+                raw_sp_name_qualifier,
+                raw_name_id
+            )
+
+        if 'attributes' in raw_subject:
+            raw_attributes = raw_subject['attributes']
+            attributes = []
+
+            for raw_attribute_name in raw_attributes:
+                raw_attribute_values = raw_attributes[raw_attribute_name]
+                attribute = Attribute(raw_attribute_name, raw_attribute_values)
+
+                attributes.append(attribute)
+
+            attribute_statement = AttributeStatement(attributes)
+
+        subject = Subject(name_id, attribute_statement)
```

Comment:
Note for myself: maybe this should produce `SubjectData` instead.

Reply:
It's a different type - `api.saml.metadata.Subject` - it contains SAML attributes and a SAML name ID. I renamed it to `SAMLSubject` in [PR # 1523](https://github.com/NYPL-Simplified/circulation/pull/1523)